Add base xDS REST SD and kuma_sd implementation

austince opened the pull request:
Implementation of a [Kuma](http://kuma.io) Service Discovery mechanism via the [xDS REST protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#rest-json-polling-subscriptions).

Since this is new code, the protobuf library used is the official `google.golang.com/protobuf` lib, which is where the community would like to migrate to, as per [this ML thread](https://groups.google.com/g/prometheus-developers/c/uFWRyqZaQis/m/Bi28HO4pAgAJ). The usage is entirely isolated from the `prompb` package where the `gogo/proto` package is currently used. When this transition is complete in Prometheus, it may make sense to include the Kuma MADS v1 `.proto` and compile it here, but since this is a "frozen" API, and the dependency on the official client already exists transitively, it did not seem too bad to include the compiled version here in the meantime.

Closes #7919

### Screenshots

![Screen Shot 2021-05-04 at 4 42 09 PM](https://user-images.githubusercontent.com/4655775/117067395-f60ed980-acf7-11eb-9956-ab2e9d8fd080.png)
Target synchronization

![Screen Shot 2021-05-04 at 4 41 44 PM](https://user-images.githubusercontent.com/4655775/117067386-f3ac7f80-acf7-11eb-81cb-739e5f4ba488.png)
Targets in Prometheus

![Screen Shot 2021-05-04 at 4 43 21 PM](https://user-images.githubusercontent.com/4655775/117067398-f7400680-acf7-11eb-8d09-cfa127f8b2c5.png)
Scraped Envoy metrics

austince commented:
I think the go version in `go.mod` is out of date (1.14) – I had to tidy with go 1.16 to pass the CI check. Did I miss this note somewhere?

roidelapluie commented:
> I think the go version in `go.mod` is out of date (1.14) – I had to tidy with go 1.16 to pass the CI check. Did I miss this note somewhere?

We did not update go.mod since we do not require specific go 1.15 or 1.16 issues.

We do not guarantee that Prometheus works properly with go 1.14 or less.
We also guarantee that some packages like TSDB can still be used with go 1.15.

austince commented:
@roidelapluie – makes sense, thanks!

roidelapluie commented on discovery/refresh/refresh.go:
```diff
@@ -15,10 +15,11 @@ package refresh
 import (
 	"context"
+	"errors"
+	"github.com/go-kit/kit/log/level"
```
this import seems misplaced now.

austince commented:
> I have started reviewing this but I have a major question first:
> 
> Why do you use refresh? Refresh is used for simple service discoveries. Fully fledged service discoveries like this should directly implement the discoverer interface and not rely on refresh. Look at consul for an example.

Thanks for taking a look! 

I used refresh as I thought it was preferred for simple polling-based SDs, like triton, docker, AWS, etc. Reading through the Consul SD now it seems like it's doing some sort of persistent watching + rpc calls on a few channels, sort of like a gRPC stream? Is that correct? 

Since the xDS HTTP API is polling based, refresh seemed applicable. Are you concerned about the applicability of the "skip refresh" functionality I've added here?

roidelapluie commented:
Consul uses http long polling, which sounds like I read from the xds doc: `Synchronous (long) polling via REST endpoints is also available for the xDS singleton APIs. `

roidelapluie commented:
```
		VersionInfo:   rc.latestVersion,
		ResponseNonce: rc.latestNonce,
```
That seems to be like the watch ID in consul.

austince commented:
> Consul uses http long polling, which sounds like I read from the xds doc: `Synchronous (long) polling via REST endpoints is also available for the xDS singleton APIs. `

That makes a lot of sense – is this is set up via the `http.Transport`? Never used long polling before. 
```
IdleConnTimeout: 2 * time.Duration(watchTimeout),
DialContext: conntrack.NewDialContextFunc(
  conntrack.DialWithTracing(),
  conntrack.DialWithName("consul_sd"),
),
```

If so, that shouldn't be a huge change.

roidelapluie commented:
Long polling means that you make an HTTP request and you get the answer when the server decides (e.g. when there is an update).

It means that we do not need to do a request every 30s, and of there is an update after 10s, we directly get the update. Then, we start another query, which might return only after 2 minutes (the next xds change).

austince commented:
Got it – ok, I'll look into if there are any changes necessary to support this on the xDS server side and then update this to use long polling. Thanks for the feedback – please let me any other questions that come up in the meantime.

austince commented:
@roidelapluie I've updated this PR to use HTTP long polling, which is available in the Kuma 1.2.0 release. Please have another look when you have a chance. Thanks!

roidelapluie commented:
I want to thank you already! I will get to it ASAP!

austince commented:
@roidelapluie friendly ping :)

roidelapluie reviewed (commented):
Yes, I have started the review. A very first look:

- You should wrap the errors in Kuma SD so it is clear that it is from kuma_sd (the errors report xds...)
- I propose we set a default in the versions (e.g. v3) or even hide it from the user. it seems an implementation detail, and we can add that parameter in the future if we need it.
- There are a couple of switch with one element that could maybe be changed to `if`?

roidelapluie commented:
note: I am not setting up my dev environment :)

austince commented:
> Yes, I have started the review. A very first look:
> 
>     * You should wrap the errors in Kuma SD so it is clear that it is from kuma_sd (the errors report xds...)
> 
>     * I propose we set a default in the versions (e.g. v3) or even hide it from the user. it seems an implementation detail, and we can add that parameter in the future if we need it.
> 
>     * There are a couple of switch with one element that could maybe be changed to `if`?

Great, thank you! Nice notes, I'll address them later today or tomorrow.


> note: I am not setting up my dev environment :)

:)

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1545,6 +1549,95 @@ for a detailed example of configuring Prometheus for Kubernetes.
+labels, following the [Prometheus label spec](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels),
+under a common prefix that is guaranteed to not be overwritten by Prometheus in future versions.
+
+* `__meta_kuma_label_{tag}`:
```
All the paragraph before can be replaced by `__meta_kuma_label_<labelname>: each label of the proxy`

roidelapluie reviewed (commented):
Thanks, I have my dev env setup, ready for a deeper look in the coming days.

austince commented:
@roidelapluie I think I have addressed all your comments, namely:
* removed unnecessary labels (versions, server)
* removed version config options so that only protocol v3 and MADS v1 are implemented
  * the few places where `ProtocolVersion` is kept internally are mostly placeholders to not completely lock the implementation in to v3, though can remove those too if you'd like 
* defaulted client_id to the hostname if not specified
* wrapped errors in kuma_sd and removed the confusing `xds_sd` prefixes

Thanks again + let me know if anything else shows up in your testing :)

austince commented:
Hey @roidelapluie, have you had time to continue with your testing?

roidelapluie commented on discovery/xds/xds_test.go:
```diff
@@ -0,0 +1,215 @@
+
+func createTestHTTPServer(t *testing.T, responder discoveryResponder) *httptest.Server {
+	return httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// validate req MIME types
```
```suggestion
		// Validate req MIME types.
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1546,6 +1550,79 @@ for a detailed example of configuring Prometheus for Kubernetes.
+
+# Configure whether HTTP requests follow HTTP 3xx redirects.
+[ follow_redirects: <bool> | default = true ]
+
```
Extra line.

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1546,6 +1550,79 @@ for a detailed example of configuring Prometheus for Kubernetes.
+# An arbitrary identifier to send to the MADS server, which should be unique to this instance.
+[ client_id: <string> | default = <host> of current machine ]
+
+## HTTP Configuration
```
We do not have this split in other SD's. I would not add it for nor,

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1546,6 +1550,79 @@ for a detailed example of configuring Prometheus for Kubernetes.
+# Address of the Kuma Control Plane's MADS xDS server.
+server: <string>
+# An arbitrary identifier to send to the MADS server, which should be unique to this instance.
+[ client_id: <string> | default = <host> of current machine ]
```
```suggestion
[ client_id: <string> | default = <hostname> ]
```

roidelapluie commented on docs/configuration/configuration.md:
````diff
@@ -1546,6 +1550,79 @@ for a detailed example of configuring Prometheus for Kubernetes.
+```yaml
+# Address of the Kuma Control Plane's MADS xDS server.
+server: <string>
+# An arbitrary identifier to send to the MADS server, which should be unique to this instance.
````
Is this really a strict requirement or could we hide it from the user? I would think that because we have the nonce it does not really have to be unique?

roidelapluie replied:
From my understamding we could remove this parameter for now and not expose it to the user.

austince replied:
Sounds good – we'll default to just `"prometheus"` if the FQDN cannot be found for some reason. This doesn't currently matter for the Kuma case, but could affect how the xDS server chooses to partition its resource cache.

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -1546,6 +1550,79 @@ for a detailed example of configuring Prometheus for Kubernetes.
+via the MADS v1 (Monitoring Assignment Discovery Service) xDS API, and will create a target for each proxy
+inside a Prometheus-enabled mesh.
+
+Available Kuma-specific meta labels:
```
Suggestion:
The following meta labels are available for each target:

roidelapluie commented on discovery/xds/xds_test.go:
```diff
@@ -0,0 +1,215 @@
+		}
+
+		if discoveryRes == nil {
+			w.WriteHeader(304)
```
return instead of else

roidelapluie commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoregistry"
```
imports are mixed here.

roidelapluie commented on discovery/xds/kuma.go:
```diff
@@ -0,0 +1,241 @@
+		},
+	}
+
+	// metrics
```
This comment is not needed.

roidelapluie replied:
This comment is not needed.

roidelapluie commented on discovery/xds/kuma.go:
```diff
@@ -0,0 +1,241 @@
+
+import (
+	"fmt"
+	"github.com/go-kit/log"
```
imports are mixed

roidelapluie commented on discovery/xds/client_test.go:
```diff
@@ -0,0 +1,163 @@
+
+	require.Equal(t, client.ResourceTypeURL(), res.TypeUrl)
+	require.Len(t, res.Resources, 0)
+	// should cache the nonce and version info
```
Then we should add this as a message to require.Equal:

require.Equal(t, "abc", client.latestNonce, "Nonce not cached")

roidelapluie commented on discovery/xds/client.go:
```diff
@@ -0,0 +1,229 @@
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
```
```suggestion
	"errors"
```

roidelapluie commented on discovery/xds/client.go:
```diff
@@ -0,0 +1,229 @@
+	client *http.Client
+	config *HTTPResourceClientConfig
+
+	// endpoint is the fully-constructed xDS HTTP endpoint
```
```suggestion
	// Endpoint is the fully-constructed xDS HTTP endpoint.
```

austince replied:
`endpoint` in lowercase is the name of the field here, should this be kept as is but add the `.` at the end?

roidelapluie commented on discovery/xds/client.go:
```diff
@@ -0,0 +1,229 @@
+
+	// endpoint is the fully-constructed xDS HTTP endpoint
+	endpoint string
+	// Caching
```
```suggestion
	// Caching.
```

roidelapluie reviewed (commented):
Here are a few reviews. I am unsure we should change the __metrics_path__, but I am aware that some kubernetes SD change the __scheme__ so I guess it is okay.

In general, I have the impression that there is a lot of debug level logging. We might want to see this down a bit.

Please verify that all comments start with a cap and end with a full stop.

austince commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+	}
+}
+
+func protoJSONUnmarshalOptions() protojson.UnmarshalOptions {
```
👍🏼 I initially did not do this as the examples I've seen always use new instances, but I've checked that the only mutation that happens during marshal/unmarshal is setting the resolver, if unset, which is not consequential here.

austince commented on discovery/xds/kuma_test.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+func mustMarshal(v proto.Message) []byte {
+	data, err := proto.Marshal(v)
```
will have to pass the error up from `getKumaMadsV1DiscoveryResponse(..)` since there is no `testing.T` in context, but that sounds reasonable to me

austince commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+		case <-ctx.Done():
+			return
+		default:
+			level.Debug(d.logger).Log("msg", "polling fetch endpoint", "source", d.source, "client", d.client.ID())
```
It was helpful to have a message each time a fetch happened for debugging, but the client ID has no value unless manually set (since it defaults to the hostname) and source currently is always `kuma`. 

I'll remove the client and source for now 👍🏼

austince commented on discovery/xds/kuma.go:
```diff
@@ -0,0 +1,241 @@
+	clientID := conf.ClientID
+	if clientID == "" {
+		var err error
+		clientID, err = os.Hostname()
```
It looks like one exists in the `scrape` package. I'll refactor that into a `util/osutil` package.

austince commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+	// metrics
+	// fetchDuration records the duration of a single fetch request, seconds
+	fetchDuration prometheus.Observer
+	// fetchSkipUpdateCount records the number of fetch requests that result in no update
```
This comment was more for future implementors using `xds.fetchDiscovery` to implement another SD, and thus adding new metrics without looking at the kuma metrics. Probably overthinking, so I'll remove this for now 👍🏼

austince commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// SDConfig is a base config for xDS-based SD mechanisms
+type SDConfig struct {
```
It may help future SDs to standardize the config options for xDS-based SDs, but probably not necessary for now.

austince replied:
I've removed the nesting indirection in KumaSD and just made it a type alias.

roidelapluie commented on util/osutil/hostname.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+// GetFqdn returns a FQDN if it's possible, otherwise falls back to hostname.
+func GetFqdn() (string, error) {
```
```suggestion
func GetFQDN() (string, error) {
```

roidelapluie commented on discovery/xds/xds_test.go:
```diff
@@ -0,0 +1,208 @@
+		SDConfig: conf,
+	}
+
+	// metrics
```
```suggestion
```

roidelapluie commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,170 @@
+
+	if response == nil {
+		// No update needed.
+		level.Debug(d.logger).Log("msg", "no update necessary")
```
that seems like a lot of debugging logs.

roidelapluie commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,170 @@
+			ticker.Stop()
+			return
+		default:
+			level.Debug(d.logger).Log("msg", "polling fetch endpoint")
```
That seems un-necessary debug logging

roidelapluie commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,170 @@
+	prometheus.MustRegister(kumaFetchDuration, kumaFetchSkipUpdateCount, kumaFetchFailuresCount)
+
+	// Register protobuf types that need to be marshalled/ unmarshalled.
+	_ = protoTypes.RegisterMessage((&v3.DiscoveryRequest{}).ProtoReflect().Type())
```
Do we need the empty assignment? if it is an error, should we deal with it?

austince replied:
True, this should be more like a `MustRegister..`

roidelapluie commented on discovery/xds/kuma.go:
```diff
@@ -0,0 +1,234 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
```
can we split github.com/prometheus/prometheus from the others?

roidelapluie commented on discovery/xds/client.go:
```diff
@@ -0,0 +1,229 @@
+		endpointURL.RawQuery = conf.ExtraQueryParams.Encode()
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, conf.Name, config.WithHTTP2Disabled(), config.WithIdleConnTimeout(conf.Timeout))
```
We could directly call NewClientFromConfig

```suggestion
	rt, err := config.NewClientFromConfig(conf.HTTPClientConfig, conf.Name, config.WithHTTP2Disabled(), config.WithIdleConnTimeout(conf.Timeout))
```

roidelapluie reviewed (commented):
Thanks,  we are getting very close

austince commented on discovery/xds/xds.go:
```diff
@@ -0,0 +1,187 @@
+	ProtocolV3 = ProtocolVersion("v3")
+)
+
+type HTTPConfig struct {
```
As in just use `config.HTTPClientConfig`? Would the suggestion then to be to put `RefreshInterval` and `FetchTimeout` directly in the `SDConfig`?

austince replied:
^ This is what I've seen in other SD implementations and what I've done.

austince commented:
Thanks for your patience and good reviews @roidelapluie – I think I've addressed all the feedback now.

roidelapluie commented:
Thanks!

austince commented:
Thank you so much @roidelapluie !
